### DISCUSSION TITLE
Designator Grouping

### DISCUSSION THREAD

**Discussion by kangaroosinaustria** [inventree/InvenTree#7791] [Ideas]
How feasible would it be to implement a designator grouping feature?

I've spend the better part of the last two weeks experimenting with deploying InvenTree to replace the now deprecated Ciiva that we were using for BOM management where I work. 
So far it looks very promising, but I notice that for a given BOM, there's no rule or reason behind whether every individual part is listed against its specific designator/reference, or whether each unique part is listed against what would ideally be a comma separated list of designators but in fact can be any arbitrary string.

In my mind it makes sense that in the back-end, each BOM item would be stored individually as its own record, and in the front-end it is just presented in either of the two methods based on a user toggle.

The drawback of simply attaching an arbitrary string that should be a comma-separated list of designators is that the database has no way of verifying the number of designators against the number of components specified in the quantity field. 

I've tried creating a BOM with only unique lines for unique parts, as well as creating a BOM where every individual part is its own line.
If checking the "Used In" page for a given part, the results reflect the difference in the notation method:
![image](https://github.com/user-attachments/assets/9c7f4f27-3de3-4f5e-9e4f-0d38676dbbcd)

![image](https://github.com/user-attachments/assets/60c94ab7-7e55-4014-9c1c-079506a7ee66)

This is how Ciiva presents:
Designator grouping turned on:
![image](https://github.com/user-attachments/assets/8eb8c16c-f6e4-4831-b0a8-4ce8d49650a0)

Designator grouping turned off:
![image](https://github.com/user-attachments/assets/04e14415-fa9b-4f53-940c-86595a6d7341)

**Comment by SchrodingersGat** [comment]
@kangaroosinaustria the BOM reference designator field is intentionally generic, so that it can be used or implemented as desired.

There are no "rules" around what can be in this field - and I do not support adding hard-coded rules around this.

However, there are options available in terms of custom validation rules (via a [plugin](https://docs.inventree.org/en/stable/extend/plugins/)). You could for example enforce a rule that the number of "comma separated designators" must match the "quantity" for the BOM item.

As for making a separate BOM line item entry for each component in an assembly (such as a common resistor) - this is not how I would recommend doing this. This is why the "quantity" field exists. But, the flexibility is there if you want to handle it this way.

When it comes to displaying the "assemblies" associated with a given part (i.e. the "Used In" table) - perhaps there is a mechanism here for grouping common assemblies when there are multiple BOM line items for a given part?

**Reply by kangaroosinaustria** [reply]
@SchrodingersGat Everything you say makes sense. Honestly, it's probably more a reflection of the poor legacy methodology we've inherited where a value change was implemented in a BOM but never updated in the schematic, rather than making the change in the design files and generating new datasets so that things actually line up.
